docs/debugging_guide.md

```diff
@@ -0,0 +1,362 @@
+
+For example, consider this crash within the `Card::is_right_bower()` function from our Euchre project.
+
+<img src="images/debug_call_stack_2.png" width="700px" />
+
+Evidently, something is very wrong with the current `Card` object. The debugger is not even able to access its `rank` or `suit` member variables to show us their values.
+
+<img src="images/debug_call_stack_3.png" width="400px" />
+
+We don't have enough information here to determine the problem, so let's click on the stack frame for `Card_less()`, which is the function that called `Card::is_right_bower()`. Now we can see the code and local variables for `Card_less()`, including the fact that the specific call that crashed was `b.is_right_bower(trump)`:
+
+<img src="images/debug_call_stack_4.png" width="700px" />
+
+That's not quite enough information, though. Evidently the card `b` is invalid, but that was a parameter to `Card_less()`. Let's keep looking for more context by again clicking the stack frame for `Simple::play_card()`.
+
+<img src="images/debug_call_stack_5.png" width="700px" />
+
+Here we finally have the information we need. We see that the parameter for `b` in this call to `Card_less()` was `hand[lowest_card_index]`, and we can also take a look at the variables window to see that `lowest_card_index` has a value of `1431823952`. This is where the bad card came from originally, which ultimately resulted in the crash deeper in `Card::is_right_bower()`.
+
+### Evaluate Expressions
+
+<div class="primer-spec-callout info icon-info" markdown="1">
+Use the **debug console** to evaluate expressions while your debugger is paused.
+</div>
+
+Let's say I'm working on a project involving matrices and I get a segfault from dereferencing `ptr` in my `Matrix_max()` function below. My debugger pauses when the segfault occurs.
+
+I can see `ptr` holds the address `0x55555586e000` and `size` is `25` in the variables window.
+
+But, I don't know how far `ptr` is from the beginning of the array. For this, I enter `ptr - mat->data` at the debug console.
+
+<img src="images/debug_feature_console_0.png" width="700px" />
+
+The debugger evaluates that pointer subtraction for me, which confirms that the pointer is far out of bounds.
+
+<img src="images/debug_feature_console_1.png" width="400px" />
+
+That suggests the loop is effectively infinite, and only stops when the pointer travels far enough to segfault. Looking back at my code (see above), I find that the loop condition `ptr < ptr + size` is incorrect.
+
+You can enter almost any valid C++ expression at the debug console - even function calls! Take a look below - we've called `Matrix_print()` from the debug console and the result shows at the terminal.
+
+<img src="images/debug_feature_console_2.png" width="800px" />
+
+### Debug Logging
```

Comment:
This is the lightest-weight strategy.  WDTY about moving it up to the first subsection?

Reply:
I think I prefer the ordering as-is, with this last. I've found many students actually experience the debugger first when they're trying to track down a crash, in which case being able to inspect program state through the debugger features itself is a first-line approach.

But also, there are a few wording things in that section that would need to be adjusted (e.g. where we contrast against the other approaches) and unless there's a strong reason to move it, it's easier to leave the ordering :).